virtcontainers/ich.go

```diff
@@ -0,0 +1,813 @@
+// Copyright (c) 2019 Ericsson Eurolab Deutschland GmbH
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package virtcontainers
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+	"fmt"
+	"net"
+	"syscall"
+
+ 
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	persistapi "github.com/kata-containers/runtime/virtcontainers/persist/api"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kata-containers/runtime/virtcontainers/store"
+	"github.com/kata-containers/runtime/virtcontainers/types"
+	"github.com/kata-containers/runtime/virtcontainers/utils"
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	
+	govmmIch "github.com/kata-containers/runtime/virtcontainers/pkg/intel/govmm/ich"
+	
+)
+
+//
+// Constants and type definitions related to cloud hypervisor
+//
+
+type ichState uint8
+
+const (
+	ichNotReady ichState = iota
+	ichReady
+)
+
+const (
+	ichTimeout            = 10
+	ichSocket             = "ich.sock"
+	vfsdSocket            = "virtiofsd.sock"
+	apiSocket             = "cloud-hypervisor.9166"
+	ichStopSandboxTimeout = 15
+)
+
+// the info struct goes to disk as hypervisor.json!
+
+type CloudHypervisorInfo struct {
+	PID int
+}
+
+//
+// cloud hypervisor state
+//
+type CloudHypervisorState struct {
+	sync.RWMutex
+	state ichState
+	PID int
+	VirtiofsdPid int
+	UUID string
+}
+
+func (s *CloudHypervisorState) set(state ichState) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.state = state
+}
+
+//
+// cloud hypervisor 
+//
+type cloudHypervisor struct {
+	id         string 
+	state      CloudHypervisorState
+	info       CloudHypervisorInfo
+	ichd       *exec.Cmd
+	store      *store.VCStore
+	config     HypervisorConfig
+	ctx        context.Context
+	socketPath string
+	ichConfig  govmmIch.Config
+	arch 	   ichArchBase
+}
+
+var ichKernelParams = []Param{
+	
+	{"root", "/dev/vda1"},
+	{"panic", "1"},
+	{"init", "/usr/lib/systemd/systemd"},
+	{"kvm-intel.nested", "1"},
+	{"no_timer_check" , ""},
+	{"noreplace-smp", ""},
+	{"page_alloc.shuffle", "1"},
```

Comment:
Might be worth adding comments for these options explaining why they are being specified.